tfx/components/example_gen/base_example_gen_executor.py

```diff
@@ -131,6 +131,47 @@ class BaseExampleGenExecutor(
   ExampleGen can override `GenerateExamplesByBeam`.
   """
 
+  def __init__(self):
+    super(BaseExampleGenExecutor, self).__init__()
+    self.input_signature = {}
+    self.output_signature = {}
+
+  def beam_io_signature(self, input_dict: Dict[Text, List[types.Artifact]],
+                        output_dict: Dict[Text, List[types.Artifact]],
+                        exec_properties: Dict[Text, Union[int, float, Text]]
+                        ) -> Tuple[Dict[Text, type], Dict[Text, type]]:
+    return self.input_signature, self.output_signature
```

Comment:
You need to derive the signatures (i.e. the span names) from the `exec_properties` here. Right now, the signatures are only populated _after_ running the Beam pipelines. This produces a chicken and egg problem that will prevent you from doing fusion (you to know the signatures before running the pipeline to perform fusion, but the signatures are only available _after_ running the pipeline).

Reply:
I derived the signatures from `exec_properties` inside `beam_io_signature()` as suggested- let me know if the new logic makes sense.